## Visión general

In [131]:
import pandas as pd

df = pd.read_csv("https://media.githubusercontent.com/media/elkofix/cardiac-risk-db/refs/heads/master/completa.csv", sep=',')

In [132]:
df.shape

(2782908, 29)

In [133]:
df.head()

11_CodEtnia_Indigena  11_CodEtnia_Negro  11_CodEtnia_Otro  \
0                   0.0                0.0               1.0   
1                   0.0                0.0               1.0   
2                   0.0                0.0               1.0   
3                   0.0                0.0               1.0   
4                   0.0                0.0               1.0   

   11_CodEtnia_Palenquero de San Basilio  11_CodEtnia_ROM (Gitano)  \
0                                    0.0                       0.0   
1                                    0.0                       0.0   
2                                    0.0                       0.0   
3                                    0.0                       0.0   
4                                    0.0                       0.0   

   11_CodEtnia_Raizal  ENDOSALUD_SI  \
0                 0.0           0.0   
1                 0.0           1.0   
2                 0.0           0.0   
3                 0.0           0.0   
4                 0.0           0.0   

   Indice Masa Corporal = Peso/talla Al Cuadrado  24_Talla  23_Peso  \
0                                      26.892323     167.0     75.0   
1                                      31.250000     160.0     80.0   
2                                      24.690405     161.0     64.0   
3                                      25.000000     160.0     64.0   
4                                      23.437500     160.0     60.0   

   25_TenArtSis  26_TenArtDitlica  36_RcbeIECA_SI  37_RcbeARA2_SI  \
0         130.0              89.0             0.0             0.0   
1         130.0              70.0             0.0             0.0   
2         160.0              80.0             0.0             0.0   
3         160.0              80.0             0.0             0.0   
4         110.0              80.0             0.0             0.0   

   22_EtiologiaERC_HTA-DM  EDAD (Años cumplidos)  8_Sexo_F  \
0                       0                   81.0         0   
1                       0                   55.0         0   
2                       0                   65.0         1   
3                       1                   78.0         1   
4                       0                   55.0         1   

   9_Regimen_CONTRIBUTIVO  9_Regimen_NO ASEGURADO  9_Regimen_SUBSIDIADO  \
0                       0                       0                     1   
1                       0                       0                     1   
2                       0                       0                     1   
3                       0                       1                     0   
4                       0                       0                     1   

   18_DxHTA_SI  18_DxHTA_NO  22_EtiologiaERC_HTA  22_EtiologiaERC_DM  \
0            1            0                    1                   0   
1            1            0                    1                   0   
2            1            0                    1                   0   
3            1            0                    0                   0   
4            0            1                    0                   1   

   CLASIFICACION FINAL DEL RIESGO_ALTO  CLASIFICACION FINAL DEL RIESGO_BAJO  \
0                                    0                                    1   
1                                    0                                    0   
2                                    0                                    0   
3                                    1                                    0   
4                                    1                                    0   

   CLASIFICACION FINAL DEL RIESGO_MODERADO  20_DxDM_SI  \
0                                        0           0   
1                                        1           0   
2                                        1           0   
3                                        0           1   
4                                        0           1   

   CLASIFICACION FINAL DEL RIESGO_MUY A

In [134]:
df.dtypes

11_CodEtnia_Indigena                             float64
11_CodEtnia_Negro                                float64
11_CodEtnia_Otro                                 float64
11_CodEtnia_Palenquero de San Basilio            float64
11_CodEtnia_ROM (Gitano)                         float64
11_CodEtnia_Raizal                               float64
ENDOSALUD_SI                                     float64
Indice Masa Corporal = Peso/talla Al Cuadrado    float64
24_Talla                                         float64
23_Peso                                          float64
25_TenArtSis                                     float64
26_TenArtDitlica                                 float64
36_RcbeIECA_SI                                   float64
37_RcbeARA2_SI                                   float64
22_EtiologiaERC_HTA-DM                             int64
EDAD (Años cumplidos)                            float64
8_Sexo_F                                           int64
9_Regimen_CONTRIBUTIVO                             int64
9_Regimen_NO ASEGURADO                             int64
9_Regimen_SUBSIDIADO                               int64
18_DxHTA_SI                                        int64
18_DxHTA_NO                                        int64
22_EtiologiaERC_HTA                                int64
22_EtiologiaERC_DM                                 int64
CLASIFICACION FINAL DEL RIESGO_ALTO                int64
CLASIFICACION FINAL DEL RIESGO_BAJO                int64
CLASIFICACION FINAL DEL RIESGO_MODERADO            int64
20_DxDM_SI                                         int64
CLASIFICACION FINAL DEL RIESGO_MUY ALTO            int64
dtype: object

En este caso identifico que hay muchas variables categóricas que estan almancenadas como int64 y float64, es significa que para cada uno de los valores estoy usando 8 bytes, sin embargo solo necesito representar unos y ceros, por lo que una opción es pasar las variables binarias a booleanos, que solo ocupan 1 byte por valor.

In [135]:
bin_cols = [col for col in df.columns if set(df[col].unique()) <= {0, 1}]
df[bin_cols] = df[bin_cols].astype('bool')

In [136]:
df.dtypes

11_CodEtnia_Indigena                                bool
11_CodEtnia_Negro                                   bool
11_CodEtnia_Otro                                    bool
11_CodEtnia_Palenquero de San Basilio               bool
11_CodEtnia_ROM (Gitano)                            bool
11_CodEtnia_Raizal                                  bool
ENDOSALUD_SI                                        bool
Indice Masa Corporal = Peso/talla Al Cuadrado    float64
24_Talla                                         float64
23_Peso                                          float64
25_TenArtSis                                     float64
26_TenArtDitlica                                 float64
36_RcbeIECA_SI                                      bool
37_RcbeARA2_SI                                      bool
22_EtiologiaERC_HTA-DM                              bool
EDAD (Años cumplidos)                            float64
8_Sexo_F                                            bool
9_Regimen_CONTRIBUTIVO                              bool
9_Regimen_NO ASEGURADO                              bool
9_Regimen_SUBSIDIADO                                bool
18_DxHTA_SI                                         bool
18_DxHTA_NO                                         bool
22_EtiologiaERC_HTA                                 bool
22_EtiologiaERC_DM                                  bool
CLASIFICACION FINAL DEL RIESGO_ALTO                 bool
CLASIFICACION FINAL DEL RIESGO_BAJO                 bool
CLASIFICACION FINAL DEL RIESGO_MODERADO             bool
20_DxDM_SI                                          bool
CLASIFICACION FINAL DEL RIESGO_MUY ALTO             bool
dtype: object

In [146]:
import pandas as pd

def resumen_dataframe(df):
    numericas = df.select_dtypes(include=['float64'])
    resumen_numericas = numericas.describe().T

    categoricas = df.select_dtypes(include=['bool'])
    valores_unicos = categoricas.nunique()

    print("Resumen de variables numéricas")
    display(resumen_numericas)
    print("\nValores únicos en categóricas/bool")
    display(valores_unicos)

    return {
        'numericas': resumen_numericas,
        'categoricas': valores_unicos
    }

resumen = resumen_dataframe(df)

Resumen de variables numéricas


count        mean  \
Indice Masa Corporal = Peso/talla Al Cuadrado  2780223.0   28.335165   
24_Talla                                       2780223.0  162.017435   
23_Peso                                        2780223.0   74.424557   
25_TenArtSis                                   2780223.0  125.393834   
26_TenArtDitlica                               2780223.0   77.877176   
EDAD (Años cumplidos)                          2780223.0   63.419833   

                                                     std    min         25%  \
Indice Masa Corporal = Peso/talla Al Cuadrado   5.438438    0.0   24.609734   
24_Talla                                        8.293579  100.0  157.000000   
23_Peso                                        15.416040    0.0   64.000000   
25_TenArtSis                                   15.508028    0.0  120.000000   
26_TenArtDitlica                                9.049822    0.0   70.000000   
EDAD (Años cumplidos)                          14.505638   18.0   54.000000   

                                                      50%     75%         max  
Indice Masa Corporal = Peso/talla Al Cuadrado   27.695133   31.25  124.700474  
24_Talla                                       162.000000  168.00  201.000000  
23_Peso                                         73.000000   83.00  258.000000  
25_TenArtSis                                   120.000000  130.00  270.000000  
26_TenArtDitlica                                80.000000   80.00  191.000000  
EDAD (Años cumplidos)                           64.000000   74.00  115.000000


Valores únicos en categóricas/bool


11_CodEtnia_Indigena                       2
11_CodEtnia_Negro                          2
11_CodEtnia_Otro                           2
11_CodEtnia_Palenquero de San Basilio      2
11_CodEtnia_ROM (Gitano)                   2
11_CodEtnia_Raizal                         2
ENDOSALUD_SI                               2
36_RcbeIECA_SI                             2
37_RcbeARA2_SI                             2
22_EtiologiaERC_HTA-DM                     2
8_Sexo_F                                   2
9_Regimen_CONTRIBUTIVO                     2
9_Regimen_NO ASEGURADO                     2
9_Regimen_SUBSIDIADO                       2
18_DxHTA_SI                                2
18_DxHTA_NO                                2
22_EtiologiaERC_HTA                        2
22_EtiologiaERC_DM                         2
CLASIFICACION FINAL DEL RIESGO_ALTO        2
CLASIFICACION FINAL DEL RIESGO_BAJO        2
CLASIFICACION FINAL DEL RIESGO_MODERADO    2
20_DxDM_SI                                 2
CLASIFICACION FINAL DEL RIESGO_MUY ALTO    2
altura_sospechosa                          1
peso_sospechoso                            2
dtype: int64

Al revisar las variables principales veo que hay varios errores o valores extremos que claramente no son realistas:  

**IMC (Indice de Masa Corporal):** la media está alrededor de 28, lo que tiene sentido para una población con sobrepeso leve. Sin embargo, el mínimo es 0 y el máximo 6500, lo cual es imposible. Probablemente hay registros mal calculados, quizá la talla estaba en cm y no se convirtió a metros. Valores realistas deberían estar entre 10 y 60.  

**Talla (cm):** la media es ~162 cm, bastante lógica. Pero hay valores mínimos de 10 cm y máximos de 260 cm. 10 cm es claramente un error y 260 cm sería un caso extraordinario (la persona más grande en colombia mide 235 cm).

**Peso (kg):** la media es ~74 kg, coherente. Sin embargo, hay registros de 0 kg (error) y hasta 258 kg, que aunque posible, es extremadamente raro. La mayoría de los casos deberían estar entre 30 y 200 kg.  

**Tensión arterial sistólica (mmHg):** media de ~125 mmHg, lógica. Min=0 (imposible) y máx=270 mmHg, extremadamente alto. Valores normales suelen estar entre 90 y 200 mmHg.  

**Tensión arterial diastólica (mmHg):** media ~78 mmHg, normal. Min=0 y máx=191 mmHg, que es casi incompatible con la vida, claramente valores atípicos.  

**Edad (años):** media ~63 años, coherente. Mínimo de 6 años sugiere que hay niños incluidos y máximo de 115 años, posible pero muy poco frecuente.  

En general, estos datos necesitan limpieza: muchos ceros, máximos imposibles y outliers que pueden distorsionar análisis posteriores.


In [138]:
num_ninos = df[df['EDAD (Años cumplidos)'] < 18].shape[0]
print("Número de niños en el dataset:", num_ninos)

num_adultos = df[df['EDAD (Años cumplidos)'] >= 18].shape[0]
print("Número de adultos en el dataset:", num_adultos)

Número de niños en el dataset: 2303
Número de adultos en el dataset: 2780605


In [139]:
df= df[df['EDAD (Años cumplidos)'] >= 18].copy()
print("Número de registros ahora:", df.shape[0])

Número de registros ahora: 2780605


Teniendo en cuenta que la participación de niños en el dataset es muy baja y que la prevalencia de riesgo cardiovascular en niños es mucho menor, con el objetivo de limpiar el dataset y evitar sesgos en el mismo, decido eliminar a la población de niños del dataset

In [ ]:
# Uso un rango arbitrario, porque al usar el rango intercuartil, me da valores completamente normales, como atípicos
# altura_col = '24_Talla'
# peso_col = '23_Peso'

# Q1_altura = df[altura_col].quantile(0.25)
# Q3_altura = df[altura_col].quantile(0.75)
# IQR_altura = Q3_altura - Q1_altura

# alturas_sospechosas = df[(df[altura_col] < Q1_altura - 1.5*IQR_altura) |
#                          (df[altura_col] > Q3_altura + 1.5*IQR_altura)]

# print(alturas_sospechosas[[altura_col, peso_col]])
# print("Número de valores atípicos:", alturas_sospechosas.shape[0])

In [140]:
def filtrar_alturas_sospechosas(df, columna_talla='24_Talla', columna_peso='23_Peso', min_talla=100, max_talla=220):
    sospechosas = df[(df[columna_talla] < min_talla) | (df[columna_talla] > max_talla)]
    print(sospechosas[[columna_talla, columna_peso]])
    print(sospechosas[[columna_talla, columna_peso]].size)
    return sospechosas, sospechosas[[columna_talla, columna_peso]].size

alturas_sospechosas_df, tamaño = filtrar_alturas_sospechosas(df)


         24_Talla  23_Peso
117083       10.0    25.00
117089       10.0     0.26
117092       10.0    25.00
117478       25.0     0.25
117580       25.0     0.25
117582       10.0     0.25
117649       10.0    25.00
117939       10.0     0.25
118020       10.0    25.00
118268       10.0    25.00
118760       10.0    25.00
118763       10.0     0.25
118862       10.0    25.00
119003       10.0     0.25
119845       10.0     0.25
119917       10.0     1.00
120262       10.0     0.25
120948       10.0     0.26
120972       10.0    25.00
121045       10.0    25.00
121451       10.0    25.00
136624       10.0    25.00
136641       10.0     0.25
149031       10.0    25.00
149249       10.0    25.00
149351       25.0     0.25
149415       10.0    25.00
149593       10.0     0.25
149645       10.0    25.00
149923       10.0    25.00
150133       10.0    25.00
150135       10.0     0.25
150204       10.0    25.00
150840       10.0     0.25
151213       10.0    25.00
152050       10.0     0.25
1

Identifico que hay filas con valores de altura de 10 y 25, y de peso de 25 y 0.25 que no tienen sentido, entonces voy a proceder a verificar si esa falta dependen de alguna otra varible, por otro lado hay otras filas en donde el peso es igual a la altura, eso es un dato faltante, sin embargo, al tener un peso razonable, puedo sumarle 100 cm a la altura, para que sea coherente

In [130]:
mask = df['23_Peso'] == df['24_Talla']
print(df[['23_Peso','24_Talla']][mask])

         23_Peso  24_Talla
16155      150.0     150.0
56172      150.0     150.0
61026      145.0     145.0
80235      163.0     163.0
117216     161.0     161.0
118205     160.0     160.0
121795     160.0     160.0
128651     157.0     157.0
137295     168.0     168.0
144011     167.0     167.0
145060     167.0     167.0
147829     165.0     165.0
151611     160.0     160.0
156844     158.0     158.0
157383     163.0     163.0
157717     222.0     222.0
165699     155.0     155.0
166126     160.0     160.0
168434     161.0     161.0
186405     165.0     165.0
186979     160.0     160.0
188895     163.0     163.0
211336     145.0     145.0
237007     165.0     165.0
267978     161.0     161.0
268967     160.0     160.0
272559     160.0     160.0
279415     157.0     157.0
288060     168.0     168.0
294776     167.0     167.0
295825     167.0     167.0
298594     165.0     165.0
302844     160.0     160.0
308619     163.0     163.0
316939     155.0     155.0
317367     160.0     160.0
3

In [141]:
columna_talla = '24_Talla'
columna_peso = '23_Peso'
imc_col = 'Indice Masa Corporal = Peso/talla Al Cuadrado'
min_talla = 100
max_talla = 220

mask_sospechosas = (df[columna_talla] < min_talla) | (df[columna_talla] > max_talla)
mask_igual_sospechosa = mask_sospechosas & (df[columna_talla] == df[columna_peso])

df.loc[mask_igual_sospechosa, columna_talla] += 100
df.loc[mask_igual_sospechosa, imc_col] = df.loc[mask_igual_sospechosa, columna_peso] / ((df.loc[mask_igual_sospechosa, columna_talla]/100)**2)

mask_no_sospechosas = ~mask_sospechosas
mask_igual_no_sospechosa = mask_no_sospechosas & (df[columna_talla] == df[columna_peso])

df.loc[mask_igual_no_sospechosa, columna_peso] -= 100
df.loc[mask_igual_no_sospechosa, imc_col] = df.loc[mask_igual_no_sospechosa, columna_peso] / ((df.loc[mask_igual_no_sospechosa, columna_talla]/100)**2)


In [143]:
import pandas as pd

df['altura_sospechosa'] = (df['24_Talla'] < 100) | (df['24_Talla'] > 220)
df['peso_sospechoso'] = (df['23_Peso'] < 30) | (df['23_Peso'] > 250)

def clase_edad(edad):
    if edad < 60:
        return 'Adulto'
    else:
        return 'Adulto Mayor'

df['clase_edad'] = df['EDAD (Años cumplidos)'].apply(clase_edad)

categoricas = [
    '11_CodEtnia_Indigena','11_CodEtnia_Negro','11_CodEtnia_Otro',
    '11_CodEtnia_Palenquero de San Basilio','11_CodEtnia_ROM (Gitano)','11_CodEtnia_Raizal',
    'ENDOSALUD_SI','36_RcbeIECA_SI','37_RcbeARA2_SI','22_EtiologiaERC_HTA-DM',
    '8_Sexo_F','9_Regimen_CONTRIBUTIVO','9_Regimen_NO ASEGURADO','9_Regimen_SUBSIDIADO',
    '18_DxHTA_SI','18_DxHTA_NO','22_EtiologiaERC_HTA','22_EtiologiaERC_DM',
    'CLASIFICACION FINAL DEL RIESGO_ALTO','CLASIFICACION FINAL DEL RIESGO_BAJO',
    'CLASIFICACION FINAL DEL RIESGO_MODERADO','20_DxDM_SI','CLASIFICACION FINAL DEL RIESGO_MUY ALTO'
]

# Crear un diccionario para guardar resultados
resultados = {}

for col in ['clase_edad'] + categoricas:
    prob_altura = df.groupby(col)['altura_sospechosa'].mean()
    prob_peso = df.groupby(col)['peso_sospechoso'].mean()
    resultados[col] = pd.DataFrame({
        'Prob_altura_sospechosa': prob_altura,
        'Prob_peso_sospechoso': prob_peso
    })

# Ejemplo: ver resultados de la edad
print(resultados['clase_edad'])

# Ejemplo: ver resultados de sexo
print(resultados['8_Sexo_F'])

for categoria, df_result in resultados.items():
    print(f"\n--- Resultados para {categoria} ---")
    print(df_result)



              Prob_altura_sospechosa  Prob_peso_sospechoso
clase_edad                                                
Adulto                      0.000183              0.000457
Adulto Mayor                0.000109              0.000247
          Prob_altura_sospechosa  Prob_peso_sospechoso
8_Sexo_F                                              
False                   0.000129              0.000225
True                    0.000141              0.000369

--- Resultados para clase_edad ---
              Prob_altura_sospechosa  Prob_peso_sospechoso
clase_edad                                                
Adulto                      0.000183              0.000457
Adulto Mayor                0.000109              0.000247

--- Resultados para 11_CodEtnia_Indigena ---
                      Prob_altura_sospechosa  Prob_peso_sospechoso
11_CodEtnia_Indigena                                              
False                               0.000135              0.000325
True                     

In [144]:
alturas_sospechosas_df, tamaño = filtrar_alturas_sospechosas(df)

         24_Talla  23_Peso
117083       10.0    25.00
117089       10.0     0.26
117092       10.0    25.00
117478       25.0     0.25
117580       25.0     0.25
117582       10.0     0.25
117649       10.0    25.00
117939       10.0     0.25
118020       10.0    25.00
118268       10.0    25.00
118760       10.0    25.00
118763       10.0     0.25
118862       10.0    25.00
119003       10.0     0.25
119845       10.0     0.25
119917       10.0     1.00
120262       10.0     0.25
120948       10.0     0.26
120972       10.0    25.00
121045       10.0    25.00
121451       10.0    25.00
136624       10.0    25.00
136641       10.0     0.25
149031       10.0    25.00
149249       10.0    25.00
149351       25.0     0.25
149415       10.0    25.00
149593       10.0     0.25
149645       10.0    25.00
149923       10.0    25.00
150133       10.0    25.00
150135       10.0     0.25
150204       10.0    25.00
150840       10.0     0.25
151213       10.0    25.00
152050       10.0     0.25
1

In [145]:
df = df.drop(alturas_sospechosas_df.index)

Los datos que nos quedan ya son errores completamente aleatorios, entonces lo que debería hacer es eliminarlos

In [147]:
resumen = resumen_dataframe(df)

Resumen de variables numéricas


count        mean  \
Indice Masa Corporal = Peso/talla Al Cuadrado  2780223.0   28.335165   
24_Talla                                       2780223.0  162.017435   
23_Peso                                        2780223.0   74.424557   
25_TenArtSis                                   2780223.0  125.393834   
26_TenArtDitlica                               2780223.0   77.877176   
EDAD (Años cumplidos)                          2780223.0   63.419833   

                                                     std    min         25%  \
Indice Masa Corporal = Peso/talla Al Cuadrado   5.438438    0.0   24.609734   
24_Talla                                        8.293579  100.0  157.000000   
23_Peso                                        15.416040    0.0   64.000000   
25_TenArtSis                                   15.508028    0.0  120.000000   
26_TenArtDitlica                                9.049822    0.0   70.000000   
EDAD (Años cumplidos)                          14.505638   18.0   54.000000   

                                                      50%     75%         max  
Indice Masa Corporal = Peso/talla Al Cuadrado   27.695133   31.25  124.700474  
24_Talla                                       162.000000  168.00  201.000000  
23_Peso                                         73.000000   83.00  258.000000  
25_TenArtSis                                   120.000000  130.00  270.000000  
26_TenArtDitlica                                80.000000   80.00  191.000000  
EDAD (Años cumplidos)                           64.000000   74.00  115.000000


Valores únicos en categóricas/bool


11_CodEtnia_Indigena                       2
11_CodEtnia_Negro                          2
11_CodEtnia_Otro                           2
11_CodEtnia_Palenquero de San Basilio      2
11_CodEtnia_ROM (Gitano)                   2
11_CodEtnia_Raizal                         2
ENDOSALUD_SI                               2
36_RcbeIECA_SI                             2
37_RcbeARA2_SI                             2
22_EtiologiaERC_HTA-DM                     2
8_Sexo_F                                   2
9_Regimen_CONTRIBUTIVO                     2
9_Regimen_NO ASEGURADO                     2
9_Regimen_SUBSIDIADO                       2
18_DxHTA_SI                                2
18_DxHTA_NO                                2
22_EtiologiaERC_HTA                        2
22_EtiologiaERC_DM                         2
CLASIFICACION FINAL DEL RIESGO_ALTO        2
CLASIFICACION FINAL DEL RIESGO_BAJO        2
CLASIFICACION FINAL DEL RIESGO_MODERADO    2
20_DxDM_SI                                 2
CLASIFICACION FINAL DEL RIESGO_MUY ALTO    2
altura_sospechosa                          1
peso_sospechoso                            2
dtype: int64